connector/connector_github.go

```diff
@@ -0,0 +1,248 @@
+)
+
+func init() {
+	RegisterConnectorConfigType(GitHubConnectorType, func() ConnectorConfig { return &GitHubConnectorConfig{} })
+}
+
+type GitHubConnectorConfig struct {
+	ID           string   `json:"id"`
+	ClientID     string   `json:"clientID"`
+	ClientSecret string   `json:"clientSecret"`
+	Scopes       []string `json:"scopes"`
+}
+
+func (cfg *GitHubConnectorConfig) ConnectorID() string {
+	return cfg.ID
+}
+
+func (cfg *GitHubConnectorConfig) ConnectorType() string {
+	return GitHubConnectorType
+}
+
+type GitHubConnector struct {
+	id           string
+	cbURL        url.URL
+	loginFunc    oidc.LoginFunc
+	clientID     string
+	clientSecret string
+	scopes       string // comma separated list of scopes
+}
+
+func (cfg *GitHubConnectorConfig) Connector(ns url.URL, lf oidc.LoginFunc, tpls *template.Template) (Connector, error) {
+	ns.Path = path.Join(ns.Path, httpPathCallback)
+	return &GitHubConnector{
+		id:           cfg.ID,
+		cbURL:        ns,
+		loginFunc:    lf,
+		clientID:     cfg.ClientID,
+		clientSecret: cfg.ClientSecret,
+		scopes:       strings.Join(cfg.Scopes, ","),
+	}, nil
+}
+
+func (c *GitHubConnector) ID() string {
+	return c.id
+}
+
+func (c *GitHubConnector) Healthy() error {
+	return nil
+}
+
+func (c *GitHubConnector) Sync() chan struct{} {
+	stop := make(chan struct{}, 1)
+	return stop
+}
+
+func (c *GitHubConnector) LoginURL(sessionKey, prompt string) (string, error) {
+	v := url.Values{}
+	v.Add("client_id", c.clientID)
+	v.Add("redirect_uri", c.cbURL.String())
+	v.Add("scope", c.scopes)
+	v.Add("state", sessionKey)
+
+	return gitHubAuthorizeURL + "?" + v.Encode(), nil
+}
+
+func (c *GitHubConnector) TrustedEmailProvider() bool {
+	return true
+}
+
+func (c *GitHubConnector) Register(mux *http.ServeMux, errorURL url.URL) {
+	log.Errorf("path %s", c.cbURL.Path)
+	mux.Handle(c.cbURL.Path, c.handleCallbackFunc(c.loginFunc, errorURL))
+}
+
+func (c *GitHubConnector) handleCallbackFunc(lf oidc.LoginFunc, errorURL url.URL) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		q := r.URL.Query()
+
+		e := q.Get("error")
+		if e != "" {
+			redirectError(w, errorURL, q)
+			return
+		}
+
+		code := q.Get("code")
+		if code == "" {
+			q.Set("error", oauth2.ErrorInvalidRequest)
+			q.Set("error_description", "code query param must be set")
+			redirectError(w, errorURL, q)
+			return
+		}
+
+		sessionKey := q.Get("state")
+		token, err := c.exchangeAuthCode(code, sessionKey)
+		if err != nil {
+			log.Errorf("Unable to verify auth code with issuer: %v", err)
+			q.Set("error", oauth2.ErrorUnsupportedResponseType)
+			q.Set("error_description", "unable to verify auth code with issuer")
+			redirectError(w, errorURL, q)
+			return
+		}
+		user, err := getGitHubUser(token)
+		if err != nil {
+			log.Errorf("Unable to query GitHub API with token: %v", err)
+			q.Set("error", oauth2.ErrorUnsupportedResponseType)
+			// TODO(eric): better error messages
```

Comment:
We need to strike a balance of not leaking information, but logging enough that an operator can tell what's going on. I think this TODO isn't necessary, unless you were thinking of something specific.